cupyx/scipy/sparse/compressed.py

```diff
@@ -550,113 +548,49 @@ def _minor_index_fancy(self, idx):
-
-    def _get_single(self, major, minor):
-        start = self.indptr[major]
-        end = self.indptr[major + 1]
-        answer = cupy.zeros((), self.dtype)
-        data = self.data[start:end]
-        indices = self.indices[start:end]
-        if self.dtype.kind == 'c':
-            self._compress_getitem_complex_kern(
-                data.real, data.imag, indices, minor, answer.real, answer.imag)
-        else:
-            self._compress_getitem_kern(
-                data, indices, minor, answer)
-        return answer[()]
-
-    def _get_submatrix(self, major=None, minor=None, copy=False):
-        """Return a submatrix of this matrix.
-        major, minor: None or slice with step 1
+    def _major_slice(self, idx, copy=False):
+        """Index along the major axis where idx is a slice object.
         """
         M, N = self._swap(*self.shape)
-        i0, i1 = self._process_slice(major, M)
-        j0, j1 = self._process_slice(minor, N)
-
-        is_major_full = i0 == 0 and i1 == M
-        is_minor_full = j0 == 0 and j1 == N
+        start, stop, step = idx.indices(M)
 
-        if is_major_full and is_minor_full:
+        if start == 0 and stop == M and step == 1:
             return self.copy() if copy else self
 
-        indptr, indices, data = self.indptr, self.indices, self.data
-
-        if not is_major_full:
-            indptr, indices, data = _index._get_csr_submatrix_major_axis(
-                indptr, indices, data, i0, i1)
-        if not is_minor_full:
-            indptr, indices, data = _index._get_csr_submatrix_minor_axis(
-                indptr, indices, data, j0, j1)
+        M = max(0, math.ceil((stop - start) / step))
+        new_shape = self._swap(M, N)
 
-        shape = self._swap(i1 - i0, j1 - j0)
-        return self.__class__((data, indices, indptr), shape=shape,
-                              dtype=self.dtype, copy=False)
+        if M == 0 or self.nnz == 0:
+            return self.__class__(new_shape)
+        if step == 1:
+            return self.__class__(
+                _index._get_csr_submatrix_major_axis(
+                    self.data, self.indices, self.indptr, start, stop),
+                shape=new_shape, copy=copy)
+        rows = cupy.arange(start, stop, step, dtype=self.indptr.dtype)
+        return self._major_index_fancy(rows)
 
-    def _major_slice(self, idx, copy=False):
-        """Index along the major axis where idx is a slice object.
+    def _minor_slice(self, idx, copy=False):
+        """Index along the minor axis where idx is a slice object.
         """
+        M, N = self._swap(*self.shape)
+        start, stop, step = idx.indices(N)
 
-        if idx == slice(None):
+        if start == 0 and stop == N and step == 1:
             return self.copy() if copy else self
 
-        M, N = self._swap(*self.shape)
-        start, stop, step = idx.indices(M)
-        M = len(range(start, stop, step))
+        N = max(0, math.ceil((stop - start) / step))
```

Comment:
`len(range(start, stop, step))` was clear and may be fast enough.  Could you explain the change?

With CPython, `length` is computed on the creation of a range object, and it uses an integer division.  The equivalent formula in Python seems
```python
if step > 0:
    lo, hi = start, stop
else:
    step = -step
    assert step
    lo, hi = stop, start
return ((hi - lo - 1) // step) + 1
```
https://github.com/python/cpython/blob/master/Objects/rangeobject.c

Reply:
I misunderstood and thought `len(range(start, stop, step))` computes by iterating range object. I will revert this fix.